### PR TITLE
feat:w3w:verify api

### PR DESCRIPTION
**Pull request by jakubuid**

### CODE REVIEW

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt [hunk 1]
```diff
@@ -68,6 +69,7 @@ fun coreAndroidNetworkModule(serverUrl: String, connectionType: ConnectionType,
     single(named(AndroidCommonDITags.OK_HTTP)) {
         OkHttpClient.Builder()
             .addInterceptor(get<Interceptor>(named(AndroidCommonDITags.INTERCEPTOR)))
+            .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/CoreNetworkModule.kt:72]: We should not add this in production. You should be able to switch to the `addInterceptor` lamda and then do a conditional check for debug builds

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/VerifyModule.kt [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+    single(named(AndroidCommonDITags.VERIFY_RETROFIT)) {
+        Retrofit.Builder()
+            .baseUrl(verifyUrl)
+            .client(get(named(AndroidCommonDITags.OK_HTTP)))
+            .addConverterFactory(MoshiConverterFactory.create(get(named(AndroidCommonDITags.MOSHI))))
+            .build()
+    }
+
+    single(named("VerifyService")) {
+        println("kobe; VerifyService")
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/VerifyModule.kt:22]: leftover log

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/VerifyModule.kt [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
+
+    single(named(AndroidCommonDITags.VERIFY_RETROFIT)) {
+        Retrofit.Builder()
+            .baseUrl(verifyUrl)
+            .client(get(named(AndroidCommonDITags.OK_HTTP)))
+            .addConverterFactory(MoshiConverterFactory.create(get(named(AndroidCommonDITags.MOSHI))))
+            .build()
+    }
+
+    single(named("VerifyService")) {
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/di/VerifyModule.kt:21]: add to tags enum

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/json_rpc/domain/JsonRpcInteractor.kt [hunk 4]
```diff
@@ -73,6 +73,9 @@ internal class JsonRpcInteractor(
         }
 
         val requestJson = serializer.serialize(payload) ?: return onFailure(IllegalStateException("JsonRpcInteractor: Unknown result params"))
+
+        println("kobe; Request: $requestJson")
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/json_rpc/domain/JsonRpcInteractor.kt:77]: log

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/json_rpc/domain/JsonRpcInteractor.kt
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/json_rpc/domain/JsonRpcInteractor.kt:1]: logs

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/VerifyClient.kt [hunk 5]
```diff
@@ -1,19 +1,49 @@
...
 import com.walletconnect.android.internal.common.di.verifyModule
+import com.walletconnect.android.internal.common.scope
+import com.walletconnect.android.internal.common.wcKoinApp
+import com.walletconnect.android.verify.data.VerifyService
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import org.koin.core.qualifier.named
 
 internal object VerifyClient : VerifyInterface {
+    private val verifyService get() = wcKoinApp.koin.get<VerifyService>(named("VerifyService"))
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/VerifyClient.kt:12]: this makes sense compared to doing lazy. Maybe we should do this for other classes like the `EchoMessagingService`?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/VerifyClient.kt [hunk 6]
```diff
@@ -1,19 +1,49 @@
...
+                            onError(IllegalArgumentException(response.body()!!.error?.message))
+                        }
+                    } else {
+                        onError(IllegalArgumentException(response.errorBody()?.string()))
+                    }
+
+                } catch (e: Exception) {
+                    onError(e)
+                }
+            }
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/VerifyClient.kt:46]: I believe we have an unwrap function that @Elyniss added that we might be able to use

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/model/VerifyServerResponse.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package com.walletconnect.android.verify.model
+
+import com.squareup.moshi.Json
+import com.squareup.moshi.JsonClass
+
+sealed class VerifyServerResponse<T> {
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/model/VerifyServerResponse.kt:6]: does this need to be a generic if both value implementations return string?

File: sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt [hunk 8]
```diff
@@ -609,6 +614,24 @@ internal class SignEngine(
                 }
             }
 
+            //todo: use usecase for that
+
+            val json = serializer.serialize(SignRpc.SessionPropose(id = request.id, params = payloadParams)) ?: throw Exception("kobe; Cannot parse")
+
+            val newJson = "\"\"\"$json\"\"\""
```
Comment (TalhaAli00) [inline sign/sdk/src/main/kotlin/com/walletconnect/sign/engine/domain/SignEngine.kt:621]: what is this supposed to do?

**Review by TalhaAli00 — APPROVED** [review]

**Review by kacperoak — APPROVED** [review]

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/storage/VerifyContextStorageRepository.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        return verifyContextQueries.geListOfVerifyContexts(mapper = this::toVerifyContext).executeAsList()
+    }
+
+    @Throws(SQLiteException::class)
+    fun delete(id: Long) {
+        verifyContextQueries.deleteVerifyContext(id)
+    }
+
+    private fun toVerifyContext(id: Long, origin: String, validation: Validation, verifyUrl: String): VerifyContext =
+        VerifyContext(id, origin, validation, verifyUrl)
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/internal/common/storage/VerifyContextStorageRepository.kt:31]: let's make all of these suspend functions

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/model/VerifyServerResponse.kt (hunk 7 above)
Comment (jakubuid) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/model/VerifyServerResponse.kt:6]: It has changed, now we have two different types

**Review by kacperoak — APPROVED** [review]

**Review by TalhaAli00 — APPROVED** [review]

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import com.walletconnect.android.internal.common.di.verifyModule
+import com.walletconnect.android.internal.common.scope
+import com.walletconnect.android.internal.common.wcKoinApp
+import com.walletconnect.android.verify.data.VerifyService
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import org.koin.core.qualifier.named
+
+internal object VerifyClient : VerifyInterface {
+    private val verifyService get() = wcKoinApp.koin.get<VerifyService>(named(AndroidCommonDITags.VERIFY_SERVICE))
```
Comment (Elyniss) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt:13]: VERIFY_SERVICE might not be necessary if you are fetching it by VerifyService interface anyway

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    override fun initialize(verifyUrl: String?) {
+        wcKoinApp.modules(verifyModule(verifyUrl))
+    }
+
+    override fun register(attestationId: String, onSuccess: () -> Unit, onError: (Throwable) -> Unit) {
+        TODO("Not yet implemented")
```
Comment (Elyniss) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt:20]: What is register supposed to do?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/domain/ResolveAttestationIdUseCase.kt [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    operator fun invoke(id: Long, jsonPayload: String, metadataUrl: String, onResolve: (VerifyContext) -> Unit) {
+        val attestationId = sha256(jsonPayload.toByteArray())
+
+        verifyInterface.resolve(attestationId,
+            onSuccess = { origin ->
+                insertContext(VerifyContext(id, origin, if (metadataUrl == origin) Validation.VALID else Validation.INVALID, verifyUrl)) { verifyContext -> onResolve(verifyContext) }
+            },
+            onError = {
+                insertContext(VerifyContext(id, "", Validation.UNKNOWN, verifyUrl)) { verifyContext -> onResolve(verifyContext) }
```
Comment (Elyniss) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/domain/ResolveAttestationIdUseCase.kt:26]: How about String.Empty here?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/domain/ResolveAttestationIdUseCase.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+import com.walletconnect.android.internal.common.storage.VerifyContextStorageRepository
+import com.walletconnect.android.internal.common.wcKoinApp
+import com.walletconnect.android.verify.client.VerifyInterface
+import com.walletconnect.android.verify.data.model.VerifyContext
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import org.koin.core.qualifier.named
+
+class ResolveAttestationIdUseCase(private val verifyInterface: VerifyInterface, private val repository: VerifyContextStorageRepository) {
+    val verifyUrl: String get() = wcKoinApp.koin.get(named(AndroidCommonDITags.VERIFY_URL))
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/domain/ResolveAttestationIdUseCase.kt:16]: can't this be moved into the constructor?

**Review by TalhaAli00 — APPROVED** [review]

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt (hunk 11 above)
Comment (jakubuid) [reply androidCore/sdk/src/main/kotlin/com/walletconnect/android/verify/client/VerifyClient.kt:20]: It's for the dapp attestation which is not a scope of this PR

**Review by Elyniss — APPROVED** [review]

**Pull request merged**